reactor-core/src/main/java/reactor/core/publisher/FluxDematerialize.java

```diff
@@ -137,79 +111,27 @@ public void onComplete() {
 				return;
 			}
 			done = true;
-			long p = produced;
-			if (p != 0L) {
-				Operators.addCap(REQUESTED, this, -p);
-			}
-			DrainUtils.postCompleteDelayError(actual, this, REQUESTED, this, this, error);
+			actual.onComplete();
 		}
 
 		@Override
 		public void request(long n) {
 			if (Operators.validate(n)) {
-				if (!DrainUtils.postCompleteRequestDelayError(n,
-						actual,
-						this,
-						REQUESTED,
-						this,
-						this,
-						error)) {
-					s.request(n);
-				}
+				s.request(n);
 			}
 		}
 
 		@Override
 		public void cancel() {
-			cancelled = true;
-			s.cancel();
+			if (!cancelled) {
```

Comment:
it is serialized, right? No need for CAS?

Reply:
no need for CAS, but some operators might be to liberal with calling `cancel()` twice (since it MUST be idempotent), hence this guard.